tianshou/data/buffer.py

```diff
@@ -181,20 +79,95 @@ def __setstate__(self, state: Dict[str, Any]) -> None:
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        """Set self.key = value."""
+        assert key not in self._reserved_keys, (
+            "key '{}' is reserved and cannot be assigned".format(key))
+        super().__setattr__(key, value)
+
+    def save_hdf5(self, path: str) -> None:
+        """Save replay buffer to HDF5 file."""
+        with h5py.File(path, "w") as f:
+            to_hdf5(self.__getstate__(), f)
+
+    @classmethod
+    def load_hdf5(
+        cls, path: str, device: Optional[str] = None
+    ) -> "ReplayBuffer":
+        """Load replay buffer from HDF5 file."""
+        with h5py.File(path, "r") as f:
+            buf = cls.__new__(cls)
+            buf.__setstate__(from_hdf5(f, device=device))
+        return buf
+
+    def reset(self) -> None:
+        """Clear all the data in replay buffer."""
+        self._index = self._size = 0
+        self._episode_length, self._episode_reward = 0, 0.0
+
+    def set_batch(self, batch: Batch) -> None:
+        """Manually choose the batch you want the ReplayBuffer to manage."""
+        assert len(batch) == self.maxsize and \
+            set(batch.keys()).issubset(self._reserved_keys), \
+            "Input batch doesn't meet ReplayBuffer's data form requirement."
+        self._meta = batch
+
+    def unfinished_index(self) -> np.ndarray:
+        """Return the index of unfinished episode."""
+        last = (self._index - 1) % self._size if self._size else 0
+        return np.array(
+            [last] if not self.done[last] and self._size else [], np.int)
+
+    def prev(self, index: Union[int, np.integer, np.ndarray]) -> np.ndarray:
+        """Return the index of previous transition.
+
+        The index won't be modified if it is the beginning of an episode.
+        """
+        index = (index - 1) % self._size
+        end_flag = self.done[index] | np.isin(index, self.unfinished_index())
+        return (index + end_flag) % self._size
+
+    def next(self, index: Union[int, np.integer, np.ndarray]) -> np.ndarray:
+        """Return the index of next transition.
+
+        The index won't be modified if it is the end of an episode.
+        """
+        end_flag = self.done[index] | np.isin(index, self.unfinished_index())
+        return (index + (1 - end_flag)) % self._size
+
+    def update(self, buffer: "ReplayBuffer") -> None:
+        """Move the data from the given buffer to current buffer."""
+        if len(buffer) == 0 or self.maxsize == 0:
+            return
+        stack_num, buffer.stack_num = buffer.stack_num, 1
+        save_only_last_obs, self._save_only_last_obs = \
+            self._save_only_last_obs, False
+        indices = buffer.sample_index(0)  # get all available indices
+        for i in indices:
+            self.add(**buffer[i])  # type: ignore
+        buffer.stack_num = stack_num
+        self._save_only_last_obs = save_only_last_obs
+
+    def alloc_fn(self, key: List[str], value: Any) -> None:
```

Comment:
I would rather define this in the constructor as a nested function by default, and stored as a private attribute.So that it is clear it is not supposed to be modified on the fly, but maybe I'm wrong about the intended use-cases.